arch/arc/core/swap.S

```diff
@@ -188,22 +191,22 @@ _swap_return_from_firq:
 	bbit1 r3, _ARC_V2_STATUS32_AE_BIT, _return_from_exc_irq
 
 	/* pretend interrupt happened to use rtie instruction */
-
-#ifdef CONFIG_ARC_HAS_SECURE
-	lr r3, [_ARC_V2_SEC_STAT]
-	/* set SEC_STAT.IRM = SECURE for interrupt return */
-	bset r3, r3, 3
-	/* sflag r3 */
-	/* sflag instruction is not supported in current ARC GNU */
-	.long 0x00ff302f
-#endif
-
+	/* use lowest interrupt priority */
 	lr r3, [_ARC_V2_AUX_IRQ_ACT]
-	brne r3, 0, _swap_already_in_irq
+#ifdef CONFIG_ARC_SECURE_FIRMWARE
+	or r3, r3, (1 << (ARC_N_IRQ_START_LEVEL - 1))
+#else
+	or r3, r3, (1 << (CONFIG_NUM_IRQ_PRIO_LEVELS - 1))
+#endif
 
-	/* use lowest interrupt priority */
-	or r3, r3, (1<<(CONFIG_NUM_IRQ_PRIO_LEVELS-1))
+#ifdef CONFIG_ARC_NORMAL_FIRMWARE
+	mov r0, _ARC_V2_AUX_IRQ_ACT
+	mov r1, r3
+	mov r6, ARC_S_CALL_AUX_WRITE
+	sjli SJLI_CALL_ARC_SECURE
```

Comment:
is it actually necessary to set IRQ_ACT before doing a 'fake' rtie? Would be good if it can be avoided.

Reply:
currently, no better way except we don't use rtie instruction. If not sec IRQ_ACT,  the rtie may call the exception return flow